### PR TITLE
Fixes for topic auto closing

### PR DESCRIPTION
**Pull request by riking**
https://meta.discourse.org/t/auto-close-topic-after-24-or-96-hours-not-working-as-expected/28900/

### CODE REVIEW

**Comment by riking** [comment]
Failures are Ember deprecations.

**Comment by SamSaffron** [comment]
@nlalonde can you review this?

File: spec/controllers/topics_controller_spec.rb [hunk 1]
```diff
@@ -882,6 +882,46 @@
         Topic.any_instance.expects(:set_auto_close).with(nil, anything)
         xhr :put, :autoclose, topic_id: @topic.id, auto_close_time: nil, auto_close_based_on_last_post: false, timezone_offset: -240
       end
+
+      it "will close a topic when the time expires" do
+        topic = Fabricate(:topic)
+        Timecop.freeze(Time.now - 20.hours) do
```
Comment (nlalonde) [inline spec/controllers/topics_controller_spec.rb]: 20.hours.ago. Also, in a rails app that's using TimeWithZone, Time.now is a Bad Idea. Always use Time.zone.now. But in this case, 20.hours.ago is best.

File: spec/controllers/topics_controller_spec.rb [hunk 2]
```diff
@@ -882,6 +882,46 @@
...
+        topic.save
+
+        Jobs.expects(:enqueue_at).at_least_once
+        xhr :put, :autoclose, topic_id: topic.id, auto_close_time: 24, auto_close_based_on_last_post: true
+
+        topic.reload
+        expect(topic.closed).to eq(false)
+        expect(topic.posts.last.raw).to match(/cool post/)
+
+        Timecop.freeze(Time.now + 5.hours) do
```
Comment (nlalonde) [inline spec/controllers/topics_controller_spec.rb]: 5.hours.from_now

File: spec/controllers/topics_controller_spec.rb [hunk 3]
```diff
@@ -882,6 +882,46 @@
...
+        end
+
+        topic.reload
+        expect(topic.closed).to eq(true)
+        expect(topic.posts.last.raw).to match(/automatically closed/)
+      end
+
+      it "will immediately close if the last post is old enough" do
+        topic = Fabricate(:topic)
+        Timecop.freeze(Time.now - 20.hours) do
```
Comment (nlalonde) [inline spec/controllers/topics_controller_spec.rb]: 20.hours.ago

**Comment by coding-horror** [comment]
@nlalonde what's the status of this?

**Comment by riking** [comment]
updated to master

**Comment by nlalonde** [comment]
Looks good, but there's a failing spec in `topic_status_update_spec.rb`. I think you meant to have this assertion on line 53:

```
expect(last_post.raw).to eq(I18n.t("topic_statuses.autoclosed_enabled_lastpost_hours", count: 10))
```

**Comment by SamSaffron** [comment]
@riking spec is still failing, this has been open for a while can we sort it out ?

**Comment by techAPJ** [comment]
Merged in via https://github.com/discourse/discourse/commit/e52852dd0f2bd2f185466d16165565bbc1f29a2c

Thanks @riking :tada:

**Pull request closed without merging**